packages/moonstone/internal/SharedStateDecorator/tests/SharedStateDecorator-specs.js

```diff
@@ -0,0 +1,307 @@
+		const Component = SharedStateDecorator(() => (
+			<SharedState.Consumer>
+				{value => {
+					value.set('key', 'value');
+
+					return value.get('key');
+				}}
+			</SharedState.Consumer>
+		));
+
+		const subject = mount(<Component />);
+
+		const expected = null;
+		const actual = subject.text();
+
+		expect(actual).toBe(expected);
+	});
+
+	test('should not set or return values when {id} is set to an empty string', () => {
+		const Component = SharedStateDecorator(() => (
+			<SharedState.Consumer>
+				{value => {
+					value.set('key', 'value');
+
+					return value.get('key');
+				}}
+			</SharedState.Consumer>
+		));
+
+		const subject = mount(<Component id="" />);
+
+		const expected = null;
+		const actual = subject.text();
+
+		expect(actual).toBe(expected);
+	});
+
+	test('should not set or return values when {id} is set to null', () => {
+		const Component = SharedStateDecorator(() => (
+			<SharedState.Consumer>
+				{value => {
+					value.set('key', 'value');
+
+					return value.get('key');
+				}}
+			</SharedState.Consumer>
+		));
+
+		const subject = mount(<Component id={null} />);
+
+		const expected = null;
+		const actual = subject.text();
+
+		expect(actual).toBe(expected);
+	});
+
+	test('should not set or return values when {id} and {noSharedState} are set', () => {
+		const Component = SharedStateDecorator(() => (
+			<SharedState.Consumer>
+				{value => {
+					value.set('key', 'value');
+
+					return value.get('key');
+				}}
+			</SharedState.Consumer>
+		));
+
+		const subject = mount(<Component id="outer" noSharedState />);
+
+		const expected = null;
+		const actual = subject.text();
+
+		expect(actual).toBe(expected);
+	});
+
+	test('should supporting deleting a value by key when {id} is set', () => {
+		const Component = SharedStateDecorator(() => (
+			<SharedState.Consumer>
+				{value => {
+					value.set('key', 'value');
+					value.delete('key');
+
+					return value.get('key');
+				}}
+			</SharedState.Consumer>
+		));
+
+		const subject = mount(<Component id="outer" />);
+
+		const expected = null;
+		const actual = subject.text();
+
+		expect(actual).toBe(expected);
+	});
+
+	test('should share data upstream when inside another SharedStateDecorator', () => {
+		const Component = SharedStateDecorator(({children, ...rest}) => (
+			<SharedState.Consumer>
+				{value => {
+					value.set('key', 'value');
+
+					return (
+						<div {...rest}>
+							<span>{value.get('key')}</span>
+							{children}
+						</div>
+					);
+				}}
+			</SharedState.Consumer>
+		));
+
+		const subject = mount(
+			<Component id="outer">
+				<Component id="inner" />
+			</Component>
+		);
+
+		const expected = {
+			// "data" is grouped by id so the top level key is the current id
+			outer: {
+				// this is set by value.set('key', 'value');
+				key: 'value',
+				// when a descendant is added, it pushes its data upstream stored by its id
+				inner: {
```

Comment:
Might be a problem if `inner` were renamed to an existing `key` value. Need to consider more.